doc/hex.1
.\" $Id: hex.1,v 1.4 2024/07/12 22:56:16 tom Exp $
.TH HEX 1 2024-07-12 MiscTools "User commands"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.de NE
.fi
.ft R
.ie n  .in -4
.el    .in -2
..
.de NS
.ie n  .sp
.el    .sp .5
.ie n  .in +4
.el    .in +2
.nf
.ft \*(CW
..
.
.hy 0
.SH NAME
hex \-
display value in different ways
.SH SYNOPSIS
.B hex
.I value
[
.I ...
]
.
.SH DESCRIPTION
Displays a value in different ways, e.g., to show ECMA codes.
.
.SH OPTIONS
None
.
.
.SH BUGS
.
Probably.
.
.
.SH EXAMPLES
.
\fBHex\fR is used to decode a string from octal, decimal or hexadecimal,
and show the value in those format as well as text or UTF-8.
For example
.
.NS 4
$ hex 1234
1234: 1234 02322 0x4d2 text "\\004\\322" utf8 \\323\\222
$ hex 0x1234
0x1234: 4660 011064 0x1234 text "\\0224" utf8 \\341\\210\\264
$ hex 01234
01234: 668 01234 0x29c text "\\002\\234" utf8 \\312\\234
.NE
.
.SH SEE ALSO
printf


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/hex/hex.c
/*
 * $Id: hex.c,v 1.5 2012/03/14 09:03:13 tom Exp $
 *
 * Title:	hex.c
 * Author:	T.E.Dickey
 * Created:	20 Feb 1995
 * Function:	displays a value in different ways, e.g., to show ECMA codes.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define EOS '\0'

static void
to_text(char *dst, unsigned long value)
{
    unsigned long p;

    dst[0] = EOS;
    for (p = value; p != 0; p >>= 8) {
	unsigned char q = (unsigned char) (p & 0xff);
	char last[80];
	strcpy(last, dst);
	if (q < 128 && isprint(q)) {
	    int m = 0;
	    dst[m++] = (char) q;
	    if (q == '\\')
		dst[m++] = (char) q;
	    dst[m] = EOS;
	} else {
	    sprintf(dst, "\\%03o", q);
	}
	strcat(dst, last);
    }
}

static void
to_utf8(char *dst, unsigned long ch)
{
    static const unsigned byteMask = 0xBF;
    static const unsigned otherMark = 0x80;
    static const unsigned firstMark[] =
    {0x00, 0x00, 0xC0, 0xE0, 0xF0, 0xF8, 0xFC};

    int result[7], *ptr;
    int count = 0;

    if ((unsigned long) ch < 0x80)
	count = 1;
    else if ((unsigned long) ch < 0x800)
	count = 2;
    else if ((unsigned long) ch < 0x10000)
	count = 3;
    else if ((unsigned long) ch < 0x200000)
	count = 4;
    else if ((unsigned long) ch < 0x4000000)
	count = 5;
    else if ((unsigned long) ch <= 0x7FFFFFFF)
	count = 6;
    else {
	count = 3;
	ch = 0xFFFD;
    }
    ptr = result + count;
    switch (count) {
    case 6:
	*--ptr = (int) ((ch | otherMark) & byteMask);
	ch >>= 6;
	/* FALLTHRU */
    case 5:
	*--ptr = (int) ((ch | otherMark) & byteMask);
	ch >>= 6;
	/* FALLTHRU */
    case 4:
	*--ptr = (int) ((ch | otherMark) & byteMask);
	ch >>= 6;
	/* FALLTHRU */
    case 3:
	*--ptr = (int) ((ch | otherMark) & byteMask);
	ch >>= 6;
	/* FALLTHRU */
    case 2:
	*--ptr = (int) ((ch | otherMark) & byteMask);
	ch >>= 6;
	/* FALLTHRU */
    case 1:
	*--ptr = (int) (ch | firstMark[count]);
	break;
    }
    while (count--) {
	sprintf(dst, "\\%03o", 0xff & *ptr++);
	dst += strlen(dst);
    }
}

int
main(int argc, char **argv)
{
    int n;

    for (n = 1; n < argc; n++) {
	char *next = argv[n];
	unsigned long value = 0;
	char *temp;
	char text_buf[BUFSIZ];
	char utf8_buf[BUFSIZ];

	do {
	    value = (value * 16) + strtoul(next, &temp, 0);
	    next = *temp ? ++temp : temp;
	} while (*next);

	to_text(text_buf, value);
	to_utf8(utf8_buf, value);
	printf("%s: %lu %#lo %#lx text \"%s\" utf8 %s\n",
	       argv[n], value, value, value, text_buf, utf8_buf);
    }
    exit(0);
}
